authentication/src/test/java/io/camunda/authentication/OidcPrivateKeyJwtKeycloakTest.java

```diff
@@ -0,0 +1,332 @@
+    final String redirectUrl = redirectResult.getResponse().getHeader("Location");
+    assertThat(redirectUrl).isNotNull();
+    final var queryParams =
+        UriComponentsBuilder.fromUriString(redirectUrl).build().getQueryParams();
+    assertThat(queryParams).containsKey("state");
+    return URLDecoder.decode(
+        Objects.requireNonNull(queryParams.getFirst("state")), StandardCharsets.UTF_8);
+  }
+
+  private void mockAuthenticatedRedirectFromIdp(final MockHttpSession session, final String state) {
+    mockMvcTester
+        .get()
+        .uri("/sso-callback")
+        .accept(MediaType.TEXT_HTML)
+        .session(session)
+        .queryParam("code", "test_authorization_code")
+        .queryParam("state", state)
+        .queryParam("session_state", "test_session_state")
+        .queryParam("iss", "http://localhost:" + wireMock.getPort() + "/realms/" + REALM)
+        .exchange();
+  }
+
+  private void verifyRequestStructure() {
+    verify(
+        1,
+        postRequestedFor(urlEqualTo(ENDPOINT_TOKEN))
+            .withHeader("Content-Type", containing("application/x-www-form-urlencoded"))
+            .withRequestBody(matching(".*grant_type=authorization_code.*"))
+            .withRequestBody(matching(".*code=test_authorization_code.*"))
+            .withRequestBody(matching(".*client_id=" + CLIENT_ID + ".*"))
+            .withRequestBody(
+                matching(
+                    ".*resource=https%3A%2F%2Fapi.example.com%2Fapp1%2F&resource=https%3A%2F%2Fapi.example.com%2Fapp2%2F.*"))
+            .withRequestBody(
+                matching(
+                    ".*client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer.*"))
+            // client_assertion=<A JWT>
+            .withRequestBody(matching(".*client_assertion=" + REGEX_JWT + ".*")));
+  }
+
+  private void verifyClientAssertion() throws ParseException {
+    final var clientAssertion = getClientAssertionFromLastRequest();
+    final var jwt = SignedJWT.parse(clientAssertion);
+
+    assertThat(jwt.getHeader().getAlgorithm()).isEqualTo(JWSAlgorithm.RS256);
+    assertThat(jwt.getHeader().getKeyID()).isEqualTo(EXPECTED_KEY_ID); // public key SHA-256
+    assertThat(jwt.getHeader().getX509CertSHA256Thumbprint()).isEqualTo(EXPECTED_X5T_S256);
+
+    // Keycloak requirements and OIDC specification match
+    // https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication
+    final Function<String, Object> getClaim = c -> jwt.getPayload().toJSONObject().get(c);
+    assertThat(getClaim.apply("iss")).isEqualTo(CLIENT_ID);
+    assertThat(getClaim.apply("sub")).isEqualTo(CLIENT_ID);
+    assertThat(getClaim.apply("aud")).isEqualTo(expectedTokenEndpointUrl);
+    assertThat(getClaim.apply("jti")).isNotNull();
+    assertThat(getClaim.apply("iat")) // issued in the past
+        .is(
+            new Condition<>(
+                iat -> Instant.now().getEpochSecond() >= (Long) iat,
```

Comment:
Here you can determine `Instant.now().getEpochSecond()` only once before the block of assertions and then reuse the variable. Same applies to the other test class.